website/addons/googledrive/tests/test_models.py

```diff
@@ -7,313 +7,370 @@
-        settings = GoogleDriveNodeSettings(user_settings=self.user_settings)
-        settings.user_settings.access_token = None
-        settings.save()
+class TestGoogleDriveNodeSettings(OsfTestCase):
 
-        assert_false(settings.has_auth)
+    def setUp(self):
+        super(TestGoogleDriveNodeSettings, self).setUp()
+        self.node = ProjectFactory()
+        self.node_settings = model.GoogleDriveNodeSettings(owner=self.node)
+        self.node_settings.save()
+        self.user = self.node.creator
+        self.user_settings = self.user.get_or_add_addon('googledrive')
+
+    def tearDown(self):
+        super(TestGoogleDriveNodeSettings, self).tearDown()
+        self.user_settings.remove()
+        self.node_settings.remove()
+        self.node.remove()
+        self.user.remove()
+
+    @mock.patch('website.addons.googledrive.model.GoogleDriveProvider')
+    def test_api_not_cached(self, mock_gdp):
+        # The first call to .api returns a new object
+        api = self.node_settings.api
+        mock_gdp.assert_called_once()
+        assert_equal(api, mock_gdp())
+
+    @mock.patch('website.addons.googledrive.model.GoogleDriveProvider')
+    def test_api_cached(self, mock_gdp):
+        # Repeated calls to .api returns the same object
+        self.node_settings._api = 'testapi'
+        api = self.node_settings.api
+        assert_false(mock_gdp.called)
+        assert_equal(api, 'testapi')
+
+    def test_set_auth(self):
+        external_account = ExternalAccountFactory()
+        self.user.external_accounts.append(external_account)
+        self.user.save()
 
-        settings.user_settings.access_token = '123abc'
-        settings.user_settings.save()
+        # this should be reset after the call
+        self.node_settings.drive_folder_id = 'anything'
 
-        assert_true(settings.has_auth)
+        self.node_settings.set_auth(
+            external_account=external_account,
+            user=self.user
+        )
 
-    # TODO use this test if delete function is used in googledrive/model
-    # def test_delete(self):
-    #     assert_true(self.node_settings.user_settings)
-    #     assert_true(self.node_settings.folder)
-    #     old_logs = self.project.logs
-    #     self.node_settings.delete()
-    #     self.node_settings.save()
-    #     assert_is(self.node_settings.user_settings, None)
-    #     assert_is(self.node_settings.folder, None)
-    #     assert_true(self.node_settings.deleted)
-    #     assert_equal(self.project.logs, old_logs)
+        # this instance is updated
+        assert_equal(
+            self.node_settings.external_account,
+            external_account
+        )
+        assert_equal(
+            self.node_settings.user_settings,
+            self.user_settings
+        )
+        assert_is_none(
+            self.node_settings.drive_folder_id
+        )
+
+        # user_settings was updated
+        # TODO: The call to grant_oauth_access in set_auth should be mocked
```

Comment:
It seems that `verify_oauth_access` uses some data set by grant- I'll let you decide whether to either act on or remove this comment.